Libraries/oneMKL/fourier_correlation/fcorr_1d_usm.cpp

```diff
@@ -13,103 +13,189 @@
 #include <mkl.h>
 #include <sycl/sycl.hpp>
 #include <iostream>
-#include <string>
-#include <oneapi/mkl/dfti.hpp>
+#include <oneapi/mkl/dft.hpp>
 #include <oneapi/mkl/rng.hpp>
 #include <oneapi/mkl/vm.hpp>
+#include <oneapi/mkl/blas.hpp>
 
-template <typename T, typename I>
-using maxloc = sycl::ext::oneapi::maximum<std::pair<T, I>>;
+template <typename T>
+static bool is_device_accessible(const T* x, const sycl::queue& Q) {
+  sycl::usm::alloc alloc_type = sycl::get_pointer_type(x, Q.get_context());
+  return (alloc_type == sycl::usm::alloc::shared
+          || alloc_type == sycl::usm::alloc::device);
+}
 
-constexpr size_t L = 1;
+static sycl::event
+naive_cross_correlation(sycl::queue& Q,
+                        unsigned int N,
+                        const float* u,
+                        const float* v,
+                        float* w,
+                        const std::vector<sycl::event> deps = {}) {
+  // u, v and w must be USM allocations of (at least) N device-accessible float
+  // values (w must be writable)
+  if (!is_device_accessible(u, Q) ||
+      !is_device_accessible(v, Q) ||
+      !is_device_accessible(w, Q)) {
+    throw std::invalid_argument("Data arrays must be device-accessible");
+  }
+  auto ev = Q.parallel_for(sycl::range<1>{N}, deps, [=](sycl::id<1> item) {
+    size_t s = item.get(0);
+    w[s] = 0.0f;
+    for (size_t j = 0; j < N; j++) {
+      w[s] += u[j] * v[(j - s + N) % N];
+    }
+  });
+  return ev;
+}
 
 int main(int argc, char** argv) {
   unsigned int N = (argc == 1) ? 32 : std::stoi(argv[1]);
-  if ((N % 2) != 0) N++;
-  if (N < 32) N = 32;
+  // N >= 8 required for the arbitrary signals as defined herein
+  if (N < 8)
+    throw std::invalid_argument("The input value N must be 8 or greater.");
```

Comment:
```suggestion
    throw std::invalid_argument("The period of the signal, chosen as input of the program, must be 8 or greater.");
```

Reply:
Done in e776e50b2647a4e7d10808cd8a9c2617005766db